proposed/0008-Connection_management.md

```diff
@@ -0,0 +1,212 @@
+- Feature Name: Improved Connection Management
+- Type: Enhancement
+- Related Components: routing, crust
+- Start Date: 23-09-2015
+- RFC PR: (leave this empty)
+- Issue Number: (leave this empty)
+
+# Summary
+
+Introduce the notion of clearly defined connection phases and establish directed routing network connections in conjunction with crust version 0.3 and later. The combined objective with crust proposed here for routing is to prepare for NAT traversal using UDP/uTP.
+
+# Motivation
+
+Starting with crust version 0.3, the API provides directional information on newly established connections.  This information is helpful for establishing a routing network with the correct topology. Already at bootstrapping it simplifies the logic for the routing node. No artificial distinction is needed for a "new bootstrap connection", as that depends on the state of the node and whether the node accepts or has connected to another node.
+
+# Detailed design
+
+## Rebootstrapping
+
+A first objective is the replacement of previous conditional variables with a clean flow through states.  This enables rebootstrapping, ie. when the connections to the network are lost, routing can automatically reconnect.  The states proposed in flow-order are `Disconnected`, `Bootstrapped`, `Relocated`, `Connected`, `GroupConnected`, `Terminated`.
+
+A full node will start at `Disconnected` and follow the states in the above order until `GroupConnected`. If a node would get disconnected then the cycle can restart at `Disconnected`.  Only when the user calls `::stop()`, does routing enter `Terminated`.  At `Terminated` the node remains active but does not accept new messages and terminates his connections to the network.
+
+A full node that cannot bootstrap (as such `Disconnected`), but accepts a connection from another node, will jump from `Disconnected` to `Relocated` by assigning itself a name.  Otherwise its behaviour is identical to normal behaviour.
+
+A client will have a reduced cycle: `Disconnected`, `Bootstrapped`, `Terminated`.  Rebootstrapping is achieved through cycling through the first two states.
+
+## improved connection management
+
+We consider first the case where the connection is not started through crust bootstrapping.  We assume that a node `A` is either already connected into the network, or has established a relay node in the network through crust bootstrapping.  
+
+## integration of Address Relocation into connection management
+
+The current mechanism of Address Relocation is compatible with the proposal here.  To activate Address Relocation a new proposal will be written that integrates Address Relocation into the `ConnectRequest`.
```

Comment:
There is no intent to update the user-events with all these internal states.  This can be discussed, but the relevant events for users are already exposed.